rust/bridge/node/futures/src/future.rs

```diff
@@ -65,100 +26,106 @@ impl<T> JsFutureState<T> {
 ///
 /// Panics in the transformation function will be propagated to the `await`ing context.
 pub struct JsFuture<T: 'static + Send> {
-    // In practice there will only be one strong reference to one of these from Rust,
-    // and two weak ones from JavaScript (for `onFulfilled` and `onRejected`).
-    state: Arc<Mutex<JsFutureState<T>>>,
+    receiver: oneshot::Receiver<std::thread::Result<T>>,
 }
 
 impl<T: 'static + Send> Future for JsFuture<T> {
     type Output = T;
 
     fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context) -> Poll<Self::Output> {
-        let mut state_guard = self.state.lock().expect("Lock can be taken");
-        let state = mem::replace(&mut *state_guard, JsFutureState::Consumed);
-        match state {
-            JsFutureState::Settled(Ok(result)) => return Poll::Ready(result),
-            JsFutureState::Settled(Err(panic)) => resume_unwind(panic),
-            JsFutureState::Consumed => panic!("already consumed"),
-            JsFutureState::Pending { .. } => {}
-        }
-        *state_guard = state.waiting_on(cx.waker().clone());
-        Poll::Pending
+        // See https://doc.rust-lang.org/std/pin/index.html#projections-and-structural-pinning.
+        let receiver = unsafe { self.map_unchecked_mut(|s| &mut s.receiver) };
+        receiver.poll(cx).map(|result| match result {
+            Ok(Ok(result)) => result,
+            Ok(Err(panic_info)) => resume_unwind(panic_info),
+            Err(_) => {
+                panic!("Future re-awaited, or JavaScript promise dropped without being settled")
+            }
+        })
     }
 }
 
-struct WeakFutureToken<T: Send + 'static>(Weak<Mutex<JsFutureState<T>>>);
+// A caught panic will never leave JsFuture in an invalid state;
+// either it's received the result already or it hasn't.
+impl<T: 'static + Send> UnwindSafe for JsFuture<T> {}
+
+/// A type describing how to settle a [JsFuture].
+///
+/// Settling a future with the result of a JavaScript promise requires
+/// 1. running the given transform synchronously to convert from a JavaScript type to a Rust type
+/// 2. forwarding the result of that (or a panic) to the waiting Rust Future.
+struct FutureSettler<T: Send + 'static> {
+    sender: oneshot::Sender<std::thread::Result<T>>,
 
-impl<T: Send + 'static> WeakFutureToken<T> {
-    /// Creates a token referencing `future`.
-    fn new(future: &JsFuture<T>) -> Self {
-        Self(Arc::downgrade(&future.state))
+    #[allow(clippy::type_complexity)]
+    transform: Box<
+        dyn for<'a> FnOnce(&mut FunctionContext<'a>, JsPromiseResult<'a>) -> T + 'static + Send,
+    >,
+}
+
+/// Represents a shared reference to a [FutureSettler].
+///
+/// Exactly one callback (from JavaScript) will be able to actually settle the future.
+// FIXME: `Mutex<Option<X>>` is heavy for "a value that can atomically only be consumed once".
+type FutureSettlerRef<T> = Arc<Mutex<Option<FutureSettler<T>>>>;
```

Comment:
I found [atomic_take](https://docs.rs/atomic-take/1.0.0/atomic_take/) which exactly what I want, but I'm not sure it's worth pulling in another crate for.